src/vs/workbench/contrib/externalTerminal/browser/externalTerminal.contribution.ts

```diff
@@ -20,119 +20,154 @@ import { IRemoteAgentService } from 'vs/workbench/services/remote/common/remoteA
 import { ContextKeyExpr } from 'vs/platform/contextkey/common/contextkey';
 import { IWorkbenchContribution, IWorkbenchContributionsRegistry, Extensions as WorkbenchExtensions } from 'vs/workbench/common/contributions';
 import { Disposable } from 'vs/base/common/lifecycle';
-import { isWeb, isWindows } from 'vs/base/common/platform';
+import { isWindows } from 'vs/base/common/platform';
 import { dirname, basename } from 'vs/base/common/path';
 import { LifecyclePhase } from 'vs/workbench/services/lifecycle/common/lifecycle';
 import { Registry } from 'vs/platform/registry/common/platform';
 import { IExternalTerminalConfiguration, IExternalTerminalService } from 'vs/platform/externalTerminal/common/externalTerminal';
 import { TerminalLocation } from 'vs/platform/terminal/common/terminal';
 
 const OPEN_IN_TERMINAL_COMMAND_ID = 'openInTerminal';
-CommandsRegistry.registerCommand({
-	id: OPEN_IN_TERMINAL_COMMAND_ID,
-	handler: async (accessor, resource: URI) => {
-		const configurationService = accessor.get(IConfigurationService);
-		const editorService = accessor.get(IEditorService);
-		const fileService = accessor.get(IFileService);
-		const integratedTerminalService = accessor.get(IIntegratedTerminalService);
-		const remoteAgentService = accessor.get(IRemoteAgentService);
-		const terminalGroupService = accessor.get(ITerminalGroupService);
-		let externalTerminalService: IExternalTerminalService | undefined = undefined;
-		try {
-			externalTerminalService = accessor.get(IExternalTerminalService);
-		} catch {
-		}
+const OPEN_IN_INTEGRATED_TERMINAL_COMMAND_ID = 'openInIntegratedTerminal';
+const OPEN_IN_EXTERNAL_TERMINAL_COMMAND_ID = 'openInExternalTerminal';
 
-		const resources = getMultiSelectedResources(resource, accessor.get(IListService), editorService, accessor.get(IExplorerService));
-		return fileService.resolveAll(resources.map(r => ({ resource: r }))).then(async stats => {
-			const targets = distinct(stats.filter(data => data.success));
-			// Always use integrated terminal when using a remote
-			const config = configurationService.getValue<IExternalTerminalConfiguration>();
-			const useIntegratedTerminal = remoteAgentService.getConnection() || config.terminal.explorerKind === 'integrated';
-			if (useIntegratedTerminal) {
-				// TODO: Use uri for cwd in createterminal
-				const opened: { [path: string]: boolean } = {};
-				const cwds = targets.map(({ stat }) => {
-					const resource = stat!.resource;
-					if (stat!.isDirectory) {
-						return resource;
-					}
-					return URI.from({
-						scheme: resource.scheme,
-						authority: resource.authority,
-						fragment: resource.fragment,
-						query: resource.query,
-						path: dirname(resource.path)
+function registerOpenTerminalCommand(id: string, explorerKind: string) {
```

Comment:
Could use a union type for explorerKind
```suggestion
function registerOpenTerminalCommand(id: string, explorerKind: 'origin' | 'integrated' | 'external') {
```